### DISCUSSION TITLE
Interest in Native iOS Version

### DISCUSSION THREAD

**Discussion by Danielku15** [CoderLine/alphaTab#1058] [Polls]
Hey alphaTabbers! 

It turns out that supporting a version of alphaTab in native iOS apps becomes and more of a challenge. This Poll shall give me a better insight if there is even interest in such a version before investing hours/days/months of work to get something working which might result in problems. Here some words on the journey: 

# Kotlin Native

Kotlin claims to be a good foundation to develop cross platform apps so we thought it would be a good choice for targeting Android and iOS native versions. So the journey on Kotlin JVM began and already it turned out much of a pain Kotlin can be. Its super strict mentality on null and thread safety delayed a lot the progress building an Android version. The performance of the tooling is comparably very poor (compiling the same codebase takes even with caching usually a lot longer than compiling JS or .net). Nevertheless we managed to build an Android version which seems to be working in its basics. 

So I started diving into Kotlin Native afterwards and we instantly got stuck because the GC of Kotlin Native had many restrictions (e.g. only thread local statics). Only lately those restrictions were lifted allowing me to progress. All in all the Kotlin tooling seems often on a alpha or beta stage only which implies a huge risk of bugs. Every time I advance one step I run into more bugs in Kotlin and Android Studio making it a huge pain to develop those versions. Some key problems currently are: 

* Apple is a closed platform and without working on Macs you cannot targeting it. I do not own a Mac and also I have no interest in spending my money to rent Macs in the cloud or buy a Mac mini plus the Apple Developer Fees. I could try to build and test a Native x64 version on desktop, and blindly compiling and testing iOS versions only through GitHub actions. Developing a full iOS UI/App integration seems out of reach. It might backfire very badly because testing/debugging of iOS version is not possible to devs. 
* The tooling of Kotlin being so poor in performance and usability compared to TS/.net makes me worry about the stability of final apps using it. 

# .net 7 AOT Compilation on the rise
 
Lately .net 7 was released and with it the amazing .net Native AoT. Being a .net developer since long time, I was eager to give it a try. In only 1 weekend (~5-6hrs of work overall) I managed to compile a native Win-x64 version of alphaTab and run it in a C++ project as Shared Library (.dll). It is lacking still the features to generate an API Layer for Libraries which can be used in apps. 

.net 7 AoT being used across the whole .net ecosystem (Xamarin/MAUI, Blazor/WASM, normal app native compilation) should give a trustworthy base for also making iOS Libraries.

**Comment by Danielku15** [comment]
Pinging some potential stakeholders. @Ltty @gburlet @zhouhao27 @kiketen

**Comment by zhouhao27** [comment]
Currently I'm trying to use Javascript to implement the score rendering in iOS built-in browser. But I have problem to pass the data (for example, tab file) from iOS to browser. I notice there are some code in `browser-advanced` example like this:

```
 const files = e.dataTransfer.files;
        if (files.length === 1) {
            const reader = new FileReader();
            reader.onload = function (data) {
                at.load(data.target.result, [0]);
            };
            reader.readAsArrayBuffer(files[0]);
        }
```

But it needs to drag the file to browser. My intention is to load raw data in native side and pass it to Javascript. Any idea how to do it?

**Reply by Danielku15** [reply]
I think you might be better set on Stackoverflow with this question. This topic is not directly related to alphaTab but a general iOS Web View topic on how to pass data back and forth. Unfortunately I do not have much experience on iOS and the WebView there but I bet there are tons of tutorials and guides on how to pass a native buffer from the App into the Web View. Your options might be: 

* Encode the byte array to a string (e.g. as base64 encoded) and use a function like `evaluateJavaScriptForData` to call a custom load function where you decode the string and pass it as arraybuffer to alphaTab. 
* Use messaging mechanisms like `pushMessage` to send data to the web view. 
* Maybe use some special app or file URIs to load data via ajax? 

AlphaTab needs simply an ArrayBuffer or Uint8Array to load the file then.
